File: backend/app/importers/plaintext.py
"""Zwykly tekst: jedna fiszka na linie, przod i tyl rozdzielone separatorem.

Format do szybkiego wklejenia listy - nie kazde zrodlo warto najpierw
przerabiac na CSV.
"""

from app.importers.base import ImportError_, ParseResult, SourceRow, suggest_mapping

COLUMN_FRONT = "przod"
COLUMN_BACK = "tyl"

#: Kolejnosc ma znaczenie - tabulator jest jednoznaczny, myslnik bywa czescia
#: tresci, wiec wymaga spacji po obu stronach.
_SEPARATORS = ["\t", " — ", " – ", " - ", " = ", " | ", ";"]

MAX_ROWS = 20_000


def _decode(data: bytes) -> str:
    for encoding in ("utf-8-sig", "cp1250", "latin-1"):
        try:
            return data.decode(encoding)
        except UnicodeDecodeError:
            continue
    raise ImportError_("Nie udalo sie rozpoznac kodowania pliku.")


def parse(data: bytes, options: dict | None = None) -> ParseResult:
    lines = [line.strip() for line in _decode(data).splitlines()]
    lines = [line for line in lines if line and not line.startswith("#")]
    if not lines:
        raise ImportError_("Plik nie zawiera zadnych linii z trescia.")

    # Wybieramy separator, ktory dzieli najwiecej linii dokladnie na dwie czesci.
    scored = {sep: sum(1 for line in lines if line.count(sep) >= 1) for sep in _SEPARATORS}
    separator = max(scored, key=lambda s: scored[s])
    if scored[separator] == 0:
        raise ImportError_(
            "Nie wykryto separatora. Oczekiwano tabulatora albo myslnika ze spacjami "
            "(np. 'kot - cat')."
        )

    warnings: list[str] = []
    rows: list[SourceRow] = []
    skipped = 0

    for line in lines[:MAX_ROWS]:
        front, found, back = line.partition(separator)
        if not found or not front.strip() or not back.strip():
            skipped += 1
            continue
        rows.append(
            SourceRow(values={COLUMN_FRONT: front.strip(), COLUMN_BACK: back.strip()})
        )

    if skipped:
        warnings.append(f"Pominieto {skipped} linii bez separatora albo z pusta strona.")
    if not rows:
        raise ImportError_("Zadna linia nie dala sie podzielic na przod i tyl.")

    columns = [COLUMN_FRONT, COLUMN_BACK]
    return ParseResult(
        source_format="text",
        columns=columns,
        rows=rows,
        suggested_mapping=suggest_mapping(columns),
        warnings=warnings,
    )

File: frontend/src/lib/local/import/plaintext.ts
/**
 * Zwykly tekst: jedna fiszka na linie, przod i tyl rozdzielone separatorem.
 * Format do szybkiego wklejenia listy.
 */

import {
  ImportParseError,
  type ParseResult,
  decodeBytes,
  sourceRow,
  suggestMapping,
} from "./base";

export const COLUMN_FRONT = "przod";
export const COLUMN_BACK = "tyl";

//: Kolejnosc ma znaczenie - tabulator jest jednoznaczny, myslnik bywa czescia
//: tresci, wiec wymaga spacji po obu stronach.
const SEPARATORS = ["\t", " — ", " – ", " - ", " = ", " | ", ";"] as const;

export const MAX_ROWS = 20_000;

export function parse(data: Uint8Array): ParseResult {
  const lines = decodeBytes(data)
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line && !line.startsWith("#"));
  if (lines.length === 0) {
    throw new ImportParseError("Plik nie zawiera zadnych linii z trescia.");
  }

  // Wybieramy separator, ktory dzieli najwiecej linii.
  let separator: string = SEPARATORS[0];
  let bestScore = -1;
  for (const candidate of SEPARATORS) {
    const score = lines.filter((line) => line.includes(candidate)).length;
    if (score > bestScore) {
      separator = candidate;
      bestScore = score;
    }
  }
  if (bestScore === 0) {
    throw new ImportParseError(
      "Nie wykryto separatora. Oczekiwano tabulatora albo myslnika ze spacjami (np. 'kot - cat').",
    );
  }

  const warnings: string[] = [];
  const rows = [];
  let skipped = 0;

  for (const line of lines.slice(0, MAX_ROWS)) {
    const at = line.indexOf(separator);
    const front = at >= 0 ? line.slice(0, at).trim() : "";
    const back = at >= 0 ? line.slice(at + separator.length).trim() : "";
    if (at < 0 || !front || !back) {
      skipped += 1;
      continue;
    }
    rows.push(sourceRow({ values: { [COLUMN_FRONT]: front, [COLUMN_BACK]: back } }));
  }

  if (skipped) {
    warnings.push(`Pominieto ${skipped} linii bez separatora albo z pusta strona.`);
  }
  if (rows.length === 0) {
    throw new ImportParseError("Zadna linia nie dala sie podzielic na przod i tyl.");
  }

  const columns = [COLUMN_FRONT, COLUMN_BACK];
  return {
    sourceFormat: "text",
    columns,
    rows,
    suggestedMapping: suggestMapping(columns),
    warnings,
    sourceDecks: [],
    suggestedNoteType: "basic",
  };
}
